File: p5_infinite_runner/eventSystem.js
class EventSystem {
    constructor() {
        this.observers = [];
    }

    addObserver(observer) {
        this.observers.push(observer);
    }

    removeObserver(observer) {
        this.observers.splice(this.observers.indexOf(observer), 1);
    }

    notify(entity, event) {
        this.observers.forEach(observer => observer.onNotify(entity, event));
    }
}

File: ggez_infinite_runner/src/event_system.rs
use super::Score;

pub enum Events {
    JUMPED_OVER_OBSTACLE,
}

pub struct EventSystem<'a> {
    observers: Vec<&'a Score>,
}

impl<'a> EventSystem<'a> {
    pub fn new() -> EventSystem<'a> {
        EventSystem { observers: vec![] }
    }

    pub fn addObserver(&mut self, observer: &'a Score) {
        self.observers.push(&observer);
    }

    pub fn notify(&self, event: Events) {
        for observer in &self.observers {
            observer.onNotify(&event);
        }
    }
}
